packages/navigator/src/browser/navigator-tree.ts

```diff
@@ -30,7 +32,102 @@ export class FileNavigatorTree extends FileTree {
+                    parent.fileStat = fileStat;
+                }
+            } else {
+                fileStat = await this.resolveFileStat(parent);
+            }
+
+            if (fileStat) {
+                resolvedChildren = await this.toNodes(fileStat, parent);
+            }
+            return this.filter.filter(resolvedChildren);
+        }
         return this.filter.filter(super.resolveChildren(parent));
     }
 
+    protected async toNode(fileStat: FileStat, parent: CompositeTreeNode): Promise<FileNode | DirNode> {
+        const id = this.generateFileTreeNodeId(fileStat.uri, parent);
+        const node = this.getNode(id);
+        if (fileStat.isDirectory) {
+            if (DirNode.is(node)) {
+                node.fileStat = fileStat;
+                return node;
+            }
+            return Object.assign(await super.toNode(fileStat, parent), { id });
+        }
+        if (FileNode.is(node)) {
+            node.fileStat = fileStat;
+            return node;
+        }
+        return Object.assign(await super.toNode(fileStat, parent), { id });
+    }
+
+    protected isVirtualRoot(node: CompositeTreeNode): boolean {
+        return !node.visible && DirNode.is(node) && !node.parent;
+    }
+
+    protected generateFileTreeNodeId(uri: string, parent: CompositeTreeNode | undefined): string {
+        if (parent) {
+            const grandparent = parent.parent;
+            if (grandparent && !this.isVirtualRoot(grandparent)) {
+                return this.generateFileTreeNodeId(uri, grandparent);
+            } else {
+                if (FileStatNode.is(parent)) {
+                    if (this.isVirtualRoot(parent)) {
+                        return `${new URI(uri).path.toString()}///${uri}`;
+                    } else {
+                        return `${parent.uri.path.toString()}///${uri}`;
+                    }
+                } else {
+                    return `${parent.name}///${uri}`;
+                }
+            }
+        }
+        return uri;
+    }
+
+    protected async resolveVirtualRootFileStat(virtualRoot: FileStatNode): Promise<FileStat | undefined> {
+        const rootFileStat = await this.fileSystem.getFileStat(virtualRoot.uri.toString());
+        const children = virtualRoot.fileStat.children;
+        let childrenFileStat: (FileStat | undefined)[] = [];
+        if (children) {
+            childrenFileStat = await Promise.all(children.map(child =>
+                this.fileSystem.getFileStat(child.uri.toString())
+            ));
+        }
+        if (rootFileStat) {
+            rootFileStat.isDirectory = true;
+            rootFileStat.children = [];
+            for (const stat of childrenFileStat) {
+                if (stat) {
+                    rootFileStat.children.push(stat);
+                }
+            }
+        }
+        return rootFileStat;
+    }
+
+}
+
+export type WorkspaceRootNode = FileStatNode & ExpandableTreeNode;
+export namespace WorkspaceRootNode {
```

Comment:
We should distinguish 2 new kinds of nodes `WorkspaceNode` and `WorkspaceRootNode`:
- `WorkspaceRootNode` should be `DirNode` with the parent of type `WorkspaceNode`
- `WorkspaceNode` should correspond to `WorkspaceDescription` and has children of `WorkspaceRootNode`